### PR TITLE
#2092 Improve physical path graph display

### LINKED ISSUE #2092: Improve physical path graph display
- Show internal links when there's a composite physical paths.
- Show all possible internal links when taking into account the fact that physical paths do not have orientation.

### PR DESCRIPTION
**Pull request by minhtutonthat**
Resolves #2092
Change-Id: I27c5752caef8e08f5f95617d57e19c214282b51c
Signed-off-by: Tu Ton <minhtutonthat@gmail.com>

### CODE REVIEW

**Comment by pdulth** [comment]
image

**Comment by pdulth** [comment]
image

File: core/plugins/org.polarsys.capella.core.sirius.analysis/src/org/polarsys/capella/core/sirius/analysis/PhysicalServices.java [hunk 1]
```diff
@@ -824,60 +829,76 @@ public void updateInternalPhysicalPaths(DDiagram diagram) {
...
-        displayedLinks.addAll(physicalPathToEdgesMap.get(physicalPath));
+      // Create and highlight internal links
+      PhysicalPathInternalLinksGraph linksGraph = PhysicalPathCache.getInstance().getInternalLinksGraph(graph);
+      for (InternalLinkEdge edge : linksGraph.getEdges().values()) {
+        PhysicalPort source = edge.getSource().getSemantic();
+        PhysicalPort target = edge.getTarget().getSemantic();
+        DDiagramElement sourceNode = DiagramServices.getDiagramServices().getDiagramElement(diagram, source);
+        DDiagramElement targetNode = DiagramServices.getDiagramServices().getDiagramElement(diagram, target);
+        if (sourceNode != null && targetNode != null) {
+          DEdge link = createInternalLink((DNode) sourceNode, (DNode) targetNode, path);
```
Comment (pdulth) [inline core/plugins/org.polarsys.capella.core.sirius.analysis/src/org/polarsys/capella/core/sirius/analysis/PhysicalServices.java:871]: As we now create internalLinks here, i think block called at L806-L809 calling updatePhysicalPathInternalLinks is no longer useful no as based on Previous/NextInvolvements? no?

**Review by pdulth — CHANGES_REQUESTED** [review]

File: core/plugins/org.polarsys.capella.core.sirius.analysis/src/org/polarsys/capella/core/sirius/analysis/PhysicalServices.java (hunk 1 above)
Comment (pdulth) [reply core/plugins/org.polarsys.capella.core.sirius.analysis/src/org/polarsys/capella/core/sirius/analysis/PhysicalServices.java:871]: I tested a bit, updatePhysicalPathInternalLinks seems useless now, if so, can you remove it and its sub method updateInternalLinks(path, map, part, list) too

**Comment by pdulth** [comment]
image

File: core/plugins/org.polarsys.capella.core.sirius.analysis/src/org/polarsys/capella/core/sirius/analysis/PhysicalServices.java (hunk 1 above)
Comment (minhtutonthat) [reply core/plugins/org.polarsys.capella.core.sirius.analysis/src/org/polarsys/capella/core/sirius/analysis/PhysicalServices.java:871]: Yes sure, I'll commit the change

File: core/plugins/org.polarsys.capella.core.sirius.analysis/src/org/polarsys/capella/core/sirius/analysis/PhysicalServices.java [hunk 2]
```diff
@@ -824,60 +819,76 @@ public void updateInternalPhysicalPaths(DDiagram diagram) {
...
       }
       
-      if (physicalPathToEdgesMap.containsKey(physicalPath)) {
-        displayedLinks.addAll(physicalPathToEdgesMap.get(physicalPath));
+      // Create and highlight internal links
+      PhysicalPathInternalLinksGraph linksGraph = PhysicalPathCache.getInstance().getInternalLinksGraph(graph);
+      for (InternalLinkEdge edge : linksGraph.getEdges().values()) {
+        PhysicalPort source = edge.getSource().getSemantic();
+        PhysicalPort target = edge.getTarget().getSemantic();
+        DDiagramElement sourceNode = DiagramServices.getDiagramServices().getDiagramElement(diagram, source);
```
Comment (pdulth) [inline core/plugins/org.polarsys.capella.core.sirius.analysis/src/org/polarsys/capella/core/sirius/analysis/PhysicalServices.java:858]: On multipart, a port might be displayed several times, maybe a 
for (sourceNodes
  for(targetNodes 
     createInternalLink
     
in createInternalLink there is a check if valid (same container node) i think

**Review by pdulth — APPROVED** [review]

**Pull request merged**